src/algebra/tsze.lean

```diff
@@ -0,0 +1,273 @@
+
+lemma inr_mul_inl [semiring R] [add_comm_monoid M] [semimodule R M] (r : R) (m : M) :
+  (inr m * inl r : tsze R M) = inr (r • m) :=
+ext (zero_mul r) $ show (0 : R) • 0 + r • m = r • m, by rw [smul_zero, zero_add]
+
+lemma inr_mul_inr [semiring R] [add_comm_monoid M] [semimodule R M] (m₁ m₂ : M) :
+  (inr m₁ * inr m₂ : tsze R M) = 0 :=
+ext (mul_zero _) $ show (0 : R) • m₂ + (0 : R) • m₁ = 0, by rw [zero_smul, zero_add, zero_smul]
+
+instance [comm_monoid R] [add_monoid M] [distrib_mul_action R M] : monoid (tsze R M) :=
+{ mul_assoc := λ x y z, ext (mul_assoc x.1 y.1 z.1) $
+    show (x.1 * y.1) • z.2 + z.1 • (x.1 • y.2 + y.1 • x.2) =
+      x.1 • (y.1 • z.2 + z.1 • y.2) + (y.1 * z.1) • x.2,
+    by simp_rw [smul_add, ← mul_smul, add_assoc, mul_comm],
+  one_mul := λ x, ext (one_mul x.1) $ show (1 : R) • x.2 + x.1 • 0 = x.2,
+    by rw [one_smul, smul_zero, add_zero],
+  mul_one := λ x, ext (mul_one x.1) $ show (x.1 • 0 : M) + (1 : R) • x.2 = x.2,
+    by rw [smul_zero, zero_add, one_smul],
+  .. triv_sq_zero_ext.has_one R M,
+  .. triv_sq_zero_ext.has_mul R M }
+
+instance [comm_semiring R] [add_comm_monoid M] [semimodule R M] : comm_semiring (tsze R M) :=
+{ mul_comm := λ x₁ x₂, ext (mul_comm x₁.1 x₂.1) $
+    show x₁.1 • x₂.2 + x₂.1 • x₁.2 = x₂.1 • x₁.2 + x₁.1 • x₂.2, from add_comm _ _,
+  zero_mul := λ x, ext (zero_mul x.1) $ show (0 : R) • x.2 + x.1 • 0 = 0,
+    by rw [zero_smul, zero_add, smul_zero],
+  mul_zero := λ x, ext (mul_zero x.1) $ show (x.1 • 0 : M) + (0 : R) • x.2 = 0,
+    by rw [smul_zero, zero_add, zero_smul],
+  left_distrib := λ x₁ x₂ x₃, ext (mul_add x₁.1 x₂.1 x₃.1) $
+    show x₁.1 • (x₂.2 + x₃.2) + (x₂.1 + x₃.1) • x₁.2 =
+      x₁.1 • x₂.2 + x₂.1 • x₁.2 + (x₁.1 • x₃.2 + x₃.1 • x₁.2),
+    by simp_rw [smul_add, add_smul, add_add_add_comm],
+  right_distrib := λ x₁ x₂ x₃, ext (add_mul x₁.1 x₂.1 x₃.1) $
+    show (x₁.1 + x₂.1) • x₃.2 + x₃.1 • (x₁.2 + x₂.2) =
+      x₁.1 • x₃.2 + x₃.1 • x₁.2 + (x₂.1 • x₃.2 + x₃.1 • x₂.2),
+    by simp_rw [add_smul, smul_add, add_add_add_comm],
+  .. triv_sq_zero_ext.monoid R M,
+  .. triv_sq_zero_ext.add_comm_monoid R M }
+
+/-- The caonnical inclusion of rings `R → tsze R M`. -/
+def inl_hom [comm_semiring R] [add_comm_monoid M] [semimodule R M] : R →+* tsze R M :=
+{ to_fun := inl,
+  map_one' := inl_one R M,
+  map_mul' := inl_mul R M,
+  map_zero' := inl_zero R M,
+  map_add' := inl_add R M }
+
+end mul
+
+section algebra
+variables (R : Type u) (M : Type v)
+
+instance [comm_semiring R] [add_comm_monoid M] [semimodule R M] : algebra R (tsze R M) :=
+{ commutes' := λ r x, mul_comm _ _,
+  smul_def' := λ r x, ext rfl $ show r • x.2 = r • x.2 + x.1 • 0, by rw [smul_zero, add_zero],
+  .. triv_sq_zero_ext.semimodule R M,
+  .. triv_sq_zero_ext.inl_hom R M }
+
+/-- The caonnical `R`-algebra projection `tsze R M → R`. -/
+def fst_hom [comm_semiring R] [add_comm_monoid M] [semimodule R M] : tsze R M →ₐ[R] R :=
```

Comment:
This can be an equiv, right? The reverse direction is `{ commutes' := sorry, ..inl_hom}`, I think?

Reply:
No, this isn't an equiv. There are maps in both directions, but they don't form a bijection.

Reply:
I need to stop making this mistake...

Perhaps the lemma I'm looking for then is something like `fst_hom.comp (algebra.of_id R _) = alg_hom.id R _`?